linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            if (ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care) != THREADAPI_OK)
+            {
+                LogError("Failure joining thread");
+            }
+        }
+
+        free(threadpool->task_array);
+        free(threadpool->thread_handle_list);
+
+        sem_destroy(&threadpool->semaphore);
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (interlocked_add(&threadpool->state, 0) != (int32_t)THREADPOOL_STATE_OPEN)
+        {
+            LogWarning("Threadpool schedule work called on non-open pool");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            //int64_t insert_pos = (interlocked_increment_64(&threadpool->insert_idx) % threadpool->task_array_count) - 1;
+            int64_t insert_pos = interlocked_exchange_64(&threadpool->insert_idx, threadpool->insert_idx+1) % threadpool->task_array_count ;
+
+            // Do we need to reallocate or do we need to roll over
+            int32_t task_state = interlocked_add(&threadpool->task_array[insert_pos].task_state, 0);
+            if (task_state != TASK_NOT_USED && task_state != TASK_RESULT_INVALID)
+            {
+                // The first task is not complete, so we just need to reallocate and
+                // go to the next item
+                if (reallocate_threadpool_array(threadpool) != 0)
+                {
+                    LogError("Failure reallocating threadpool");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    insert_pos = interlocked_add_64(&threadpool->insert_idx, 0) - 1;
+                    result = 0;
+                }
+            }
+            else
+            {
+                result = 0;
+            }
+
+            if (result == 0)
+            {
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                THREADPOOL_TASK* task_item = &threadpool->task_array[insert_pos];
+                task_item->task_param = work_function_context;
+                task_item->task_func = work_function;
+                (void)interlocked_exchange(&task_item->task_state, TASK_WAITING);
```

Comment:
>)interlocked_exchange(&task_item->task_state, TASK_WAITING); [](http://example.com/codeflow?start=21&length=60)

I think this should be done with interlocked_compare_exchange to be fully correct.
It can happen that 2 producing threads get the same entry in the array if a thread is EXTREMELLY lazy and the array gets overrun by other producing/consuming threads.